py/terra_widgets/html_copies.py

```diff
@@ -0,0 +1,348 @@
+      if items:
+        time_chooser.options = collections.OrderedDict(sorted(
+            {os.path.basename(item): item for item in items}.items(), reverse=True))
+  date_chooser.observe(on_choose_date, names='value')
+
+  def on_choose_time(changed):
+    file_chooser.options = []
+    if changed['new']:
+      workspace_paths = ws_paths[workspace_chooser.value]
+      items = tf.io.gfile.glob(pattern=workspace_paths.add_html_glob_to_path(path=changed['new']))
+      if items:
+        file_chooser.options = {os.path.basename(item): item for item in items}
+  time_chooser.observe(on_choose_time, names='value')
+
+  return widgets.VBox(
+      [widgets.HTML('<h3>View an HTML copy of a notebook</h3>'), workspace_chooser, user_chooser,
+       date_chooser, time_chooser, file_chooser, widgets.HBox([view_comment_button, view_html_button])],
+      layout=widgets.Layout(width='auto', border='solid 1px grey'))
+
+
+def view_all_comments_ui(ws_names2id: Dict[str, str], ws_paths: Dict[str, WorkspacePaths], output):
+  """Create an ipywidget UI to display the contents of all comment files within a particular workspace."""
+  workspace_chooser = widgets.Dropdown(
+      options=ws_names2id,
+      value=None,
+      description='Choose a workspace to view:',
+      style={'description_width': 'initial'},
+      layout=widgets.Layout(width='900px')
+  )
+
+  def on_choose_workspace(changed):
+    with output:
+      output.clear_output()
+      workspace_paths = ws_paths[changed['new']]
+      comment_files = get_ipython().getoutput(f'gsutil ls {workspace_paths.get_comment_file_glob()}')
+      if not comment_files[0].startswith('gs://'):
+        display(HTML('''<div class="alert alert-block alert-warning">
+          No comment files found for HTML copies in this workspace.</div>'''))
+        return
+      progress = widgets.IntProgress(
+          value=0,
+          min=0,
+          max=len(comment_files),
+          step=1,
+          description=f'Retrieving {len(comment_files)} comments:',
+          bar_style='success',
+          orientation='horizontal',
+          layout=widgets.Layout(width='450px'),
+          style={'description_width': 'initial'}
+      )
+      display(progress)
+      comment_num = 0
+      comment_file_contents = []
+      for file in comment_files:
+        comment = get_ipython().getoutput(f"gsutil cat '{file}'")
+        version = file.replace(workspace_paths.get_subfolder(), '')
+        comment_file_contents.append({'version': version, 'comment': comment})
+        comment_num += 1
+        progress.value = comment_num
+      comments = pd.DataFrame(comment_file_contents)
+      display(comments)
+  workspace_chooser.observe(on_choose_workspace, names='value')
+
+  return widgets.VBox(
+      [widgets.HTML('<h3>View all comments for a workspace</h3>'), workspace_chooser],
+      layout=widgets.Layout(width='auto', border='solid 1px grey'))
+
+
+def html_copies_ui():
```

Comment:
I think this function name could be made clearer.  "copies" is slightly ambiguous as either a verb or noun, and "html" is also slightly ambiguous as to whether it refers to notebooks, or the widget itself. There's also no verb in the function name, but maybe that's standard for widgets, e.g. "render_",  "show_", "display_". Is it also worth specifying it's rendering a widget, rather than a generic "ui"?

Also, rather than calling these "copies", should they be called something like "snapshots", or "revisions" instead? This may make the intent clearer.

Ideas:
- nb_to_html_ui() ?
- display_notebook_html_snapshots_widget()?

Reply:
renamed! I think it's better now. Happy to make more changes.